sound/soc/sof/intel/hda-dai.c

```diff
@@ -165,6 +165,30 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
 	return hda_link_dma_params(link_dev, &p_params);
 }
 
+static int hda_link_pcm_prepare(struct snd_pcm_substream *substream,
+				struct snd_soc_dai *dai)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_sof_dev *sdev =
+		snd_soc_component_get_drvdata(dai->component);
+	struct snd_sof_pcm *spcm;
+	int stream = substream->stream;
+
+	spcm = snd_sof_find_spcm_dai(sdev, rtd);
+	if (!spcm)
+		return -EINVAL;
+
+	/* do hw_params only if resuming from system suspend */
+	if (!spcm->hw_params_upon_resume[stream])
+		return 0;
+
```

Comment:
prepare should handle both resume and non-resume operation. This is what I talked with you last time I want to move some operations into prepare. Prepare is only for resume seems not be very good.

Reply:
@libinyang im not sure what non-resume ops you're talking about. EVen if that is the case, its beyond the scope of this PR.

Reply:
@ranj063 Prepare is not used only for resume. Actually we do some operations in other callbacks which should be in prepare. As prepare callback is added, we should make the prepare be more general. Otherwise, it's more like a workaround for resume.